### PR TITLE
fix: account linking tests

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

Tests for:
- GetUserById (function and API)

## Related issues

- Link to issue1 here
- Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your
changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here
highlighting the necessary changes)

## Checklist for important updates

- [ ] Changelog has been updated
    - [ ] If there are any db schema changes, mention those changes clearly
- [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [ ] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] Changes to the version if needed
    - In `build.gradle`
- [ ] If added a new paid feature, edit the `getPaidFeatureStats` function in FeatureFlag.java file
- [ ] Had installed and ran the pre-commit hook
- [ ] If there are new dependencies that have been added in `build.gradle`, please make sure to add them
  in `implementationDependencies.json`.
- [ ] Update function `getValidFields` in `io/supertokens/config/CoreConfig.java` if new aliases were added for any core config (similar to the `access_token_signing_key_update_interval` config alias).
- [ ] Issue this PR against the latest non released version branch.
    - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the
      latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR

- [ ] Item1
- [ ] Item2

### CODE REVIEW

File: src/test/java/io/supertokens/test/accountlinking/GetUserByAccountInfoTest.java [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+        AuthRecipeUserInfo user4 = createPasswordlessUserWithPhone(process.getProcess(), "+919876543210");
+
+        TenantIdentifierWithStorage tenantIdentifierWithStorage = TenantIdentifier.BASE_TENANT.withStorage(StorageLayer.getBaseStorage(process.getProcess()));
+
+        // test for result
+        assertEquals(user1, AuthRecipe.getUsersByAccountInfo(tenantIdentifierWithStorage, false, "test1@example.com", null, null, null)[0]);
+        assertEquals(user2, AuthRecipe.getUsersByAccountInfo(tenantIdentifierWithStorage, false, null, null, "google", "userid1")[0]);
+        assertEquals(user2, AuthRecipe.getUsersByAccountInfo(tenantIdentifierWithStorage, false, "test2@example.com", null, "google", "userid1")[0]);
+        assertEquals(user3, AuthRecipe.getUsersByAccountInfo(tenantIdentifierWithStorage, false, "test3@example.com", null, null, null)[0]);
+        assertEquals(user4, AuthRecipe.getUsersByAccountInfo(tenantIdentifierWithStorage, false, null, "+919876543210", null, null)[0]);
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/accountlinking/GetUserByAccountInfoTest.java:118]: you also should test the structure of the output for at least one of these.

File: src/test/java/io/supertokens/test/accountlinking/GetUserByAccountInfoTest.java [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+            assertTrue(Arrays.asList(users).contains(user2));
+        }
+        {
+            AuthRecipeUserInfo[] users = AuthRecipe.getUsersByAccountInfo(tenantIdentifierWithStorage, true, "test1@example.com", "+919876543210", "google", "userid1");
+            assertEquals(3, users.length);
+            assertTrue(Arrays.asList(users).contains(user1));
+            assertTrue(Arrays.asList(users).contains(user2));
+            assertTrue(Arrays.asList(users).contains(user4));
+        }
+
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/accountlinking/GetUserByAccountInfoTest.java:173]: need to test with false option too?

File: src/test/java/io/supertokens/test/accountlinking/GetUserByIdTest.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+        assertFalse(user1.isPrimaryUser);
+        assertFalse(user2.isPrimaryUser);
+        assertFalse(user3.isPrimaryUser);
+        assertFalse(user4.isPrimaryUser);
+
+        assertEquals(user1, AuthRecipe.getUserById(process.getProcess(), user1.id));
+        assertEquals(user2, AuthRecipe.getUserById(process.getProcess(), user2.id));
+        assertEquals(user3, AuthRecipe.getUserById(process.getProcess(), user3.id));
+        assertEquals(user4, AuthRecipe.getUserById(process.getProcess(), user4.id));
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/accountlinking/GetUserByIdTest.java:118]: test structure of at least one of these.

File: src/test/java/io/supertokens/test/accountlinking/GetUserByIdTest.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            assertEquals(user2.loginMethods[0], result.loginMethods[1]);
+            assertEquals(user3.loginMethods[0], result.loginMethods[2]);
+            assertEquals(user4.loginMethods[0], result.loginMethods[3]);
+
+            assertEquals(user1.timeJoined, result.timeJoined);
+        }
+
+        process.kill();
+        assertNotNull(process.checkOrWaitForEvent(ProcessState.PROCESS_STATE.STOPPED));
+    }
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/accountlinking/GetUserByIdTest.java:140]: - need to test that this function returns null if user does not exist
- need to test that the order of login methods is sorted in the correct way (if you remove the sorting logic, the test should fail - make sure of that)

File: src/test/java/io/supertokens/test/accountlinking/api/GetUserByAccountInfoTest.java [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+        assertEquals(0, getUsersByAccountInfo(process.getProcess(), false, "test1@example.com", "+919876543210", null, null).size());
+        assertEquals(0, getUsersByAccountInfo(process.getProcess(), false, "test2@example.com", "+919876543210", null, null).size());
+        assertEquals(0, getUsersByAccountInfo(process.getProcess(), false, "test3@example.com", "+919876543210", null, null).size());
+        assertEquals(0, getUsersByAccountInfo(process.getProcess(), false, null, "+919876543210", "google", "userid1").size());
+        assertEquals(0, getUsersByAccountInfo(process.getProcess(), false, "test1@gmail.com", null, "google", "userid1").size());
+        assertEquals(0, getUsersByAccountInfo(process.getProcess(), false, "test3@gmail.com", null, "google", "userid1").size());
+
+        process.kill();
+        assertNotNull(process.checkOrWaitForEvent(ProcessState.PROCESS_STATE.STOPPED));
+    }
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/accountlinking/api/GetUserByAccountInfoTest.java:170]: test with user id mapping as well - make sure that the returned user object has the user id mapped correctly.

File: src/test/java/io/supertokens/test/accountlinking/api/GetUserByIdTest.java [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+                phoneNumbers.add(phoneNumberElem.getAsString());
+            }
+            assertEquals(2, phoneNumbers.size());
+            assertTrue(phoneNumbers.contains("+919876543210"));
+            assertTrue(phoneNumbers.contains("+911234567890"));
+        }
+
+        process.kill();
+        assertNotNull(process.checkOrWaitForEvent(ProcessState.PROCESS_STATE.STOPPED));
+    }
```
Comment (rishabhpoddar) [inline src/test/java/io/supertokens/test/accountlinking/api/GetUserByIdTest.java:434]: test unkonwn user id
- test with user id mapping as well - make sure that the returned user object has the user id mapped correctly, and - that the input user id has to be mapped, or unmapped (both should return the right user object)

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/test/java/io/supertokens/test/accountlinking/GetUserByAccountInfoTest.java (hunk 1 above)
Comment (sattvikc) [reply src/test/java/io/supertokens/test/accountlinking/GetUserByAccountInfoTest.java:118]: Done

File: src/test/java/io/supertokens/test/accountlinking/GetUserByAccountInfoTest.java (hunk 2 above)
Comment (sattvikc) [reply src/test/java/io/supertokens/test/accountlinking/GetUserByAccountInfoTest.java:173]: this test is about union set to true. there is a different test that tests with false value

File: src/test/java/io/supertokens/test/accountlinking/GetUserByIdTest.java (hunk 3 above)
Comment (sattvikc) [reply src/test/java/io/supertokens/test/accountlinking/GetUserByIdTest.java:118]: Done

File: src/test/java/io/supertokens/test/accountlinking/api/GetUserByAccountInfoTest.java (hunk 5 above)
Comment (sattvikc) [reply src/test/java/io/supertokens/test/accountlinking/api/GetUserByAccountInfoTest.java:170]: Done

File: src/test/java/io/supertokens/test/accountlinking/GetUserByIdTest.java (hunk 4 above)
Comment (sattvikc) [reply src/test/java/io/supertokens/test/accountlinking/GetUserByIdTest.java:153]: Done

File: src/test/java/io/supertokens/test/accountlinking/api/GetUserByIdTest.java (hunk 6 above)
Comment (sattvikc) [reply src/test/java/io/supertokens/test/accountlinking/api/GetUserByIdTest.java:434]: Done
